### PR TITLE
Feature: Bulk Destroy

### PR DESCRIPTION
**Pull request by jacobdaddario**
Hey! I'm super happy I managed to get this mostly done today.

This completes the feature request for bulk destroy. I'm really happy with how everything came out.

There's two things that require feedback before merging this commit:

- [ ] This feature needs tests to cover its functionality
- [ ] Deleting with nothing selected totally works when tested in the browser, but reading some Rails source, I'm surprised it doesn't throw an error given the use of `find` in the `destroy` class method
- [ ] Are the admin authorization checks in front of the madmin namespace by default or do I need to add one?

We can go ahead and merge now if you're okay with doing the tests in a different PR and trusting that the lack of error in the app when passing an empty destroy hash is okay. Otherwise I'll have to return to this later. I've done none of my chores today and need to get to them so I'm not kicking myself when I have next Friday off of work.

Regardless, I had a blast doing this. I got to practice writing bug-fixes for #105 and #107 and cherry-picking for #107. Plus, I got to do some interactive rebasing to create some semantic commits. That's all stuff I can't do when working alone, so thanks for being active on this repository.

### CODE REVIEW

**Comment by jacobdaddario** [comment]
Force pushed to split something out of the counter commit that should have been in the server logic commit.

**Comment by jacobdaddario** [comment]
Thinking some more on this, do we need to add a route to the install generator? I just checked and didn't see anything for the dashboard, so I'm wondering if we don't need anything?

**Comment by jespr** [comment]
How would you change this, if in the future you wanted to add something like "Mark as spam" or similar bulk actions?

**Comment by jespr** [comment]
Also, I think I would have appreciated a comment on whatever was faulty in my approach on my PR - or a "mind if I take this over?" instead of just seeing  a new PR being opened and not a word about it in mine. But real like took over, and I snoozed on my approach.. so it definitely sat longer than it should have. Still would have appreciated a heads up, as I could've easily continued working on my PR without ever going through the list of open PRs on the repo. It looks like you might have referenced my PR at some point - but that might've been edited away from the description as I can only see it on #89 and not here.

**Comment by jespr** [comment]
I like your approach better. If the view comes with the bulk actions per default, I'd say you need to have the route addition in the install generator as well - as functionality will be broken without

**Comment by jacobdaddario** [comment]
Hey! Sorry for taking a bit to reply to this. I saw a notification on my phone, but I only just now had an opportunity to take a break from work in order to reply to this.

> Also, I think I would have appreciated a comment on whatever was faulty in my approach on my PR - or a "mind if I take this over?" instead of just seeing a new PR being opened and not a word about it in mine. But real like took over, and I snoozed on my approach.. so it definitely sat longer than it should have. Still would have appreciated a heads up, as I could've easily continued working on my PR without ever going through the list of open PRs on the repo. It looks like you might have referenced my PR at some point - but that might've been edited away from the description as I can only see it on #89 and not here.

Sorry for that! I didn't mean any offense. So some background to this: originally, I pulled down your branch and rebased it on top of the current `master` branch. I poked around a little bit and found that I was confused about some of the action items that were left on the WIP PR. Additionally, the way Rails does destruction for nested forms came to mind along with [Matt Swanson's excellent `checkbox-list` controller](https://boringrails.com/articles/better-stimulus-controllers/), and I felt that I could take things a different direction. This would allow us to use a normal form to complete everything instead.

Additionally I picked up the PR since I thought it had been marked WIP with the intent that someone else might pick it up when they had time. I should have reached out to you about that first.

I had originally written some of this in the PR which is why you got a notification. Around the time I was going to bed, a maintainer reached out to me and expressed that they felt that it had come across as too-strongly worded. I agreed that what was written didn't align with my intent, but in my haste to get to bed, I opted to just delete the section entirely. Here's the original second paragraph of the PR message:

> This completes the feature request for bulk destroy. I tried checking out #89, and I ultimately found I wasn't totally happy with how Jespr was handling sending requests. Ultimately I ripped out the guts of his commit and combined it with a customization of Matt Swanson's checkbox-list controller. I'm really happy with how everything came out, and I feel that it's a bit simpler than the request.js logic used in the other branch. Additionally, this benefits from using some more-semantic markup by using form tags.

I should've taken the time to properly edit the content of that paragraph. I hope that there's no hard feelings for this. I program purely as a hobby and enjoy working on this project because it enables me to work collaboratively and ship small amounts of code. In the spirit of collaboration I'd never want to step on someone's toes.

Finally:
> How would you change this, if in the future you wanted to add something like "Mark as spam" or similar bulk actions?

I pulled some of the logic you originally had in the controller into a model called `Bulk`. Even though that was probably a bit of premature design, my hope is that the bulk data model would provide an interface for interacting with other kinds of bulk operations in the future.

I hope that addressed your questions and concerns. Sorry again for the misunderstanding. I'll be able to check this PR out again after work today, but until then I need to get back to things so that I don't end up leaving too late in the day.

**Comment by jacobdaddario** [comment]
I found where the routes were being auto generated and added two lines in so that the bulk destroy route is automatically added to a project. Tested it in the browser and works well.

**Comment by excid3** [comment]
Existing Madmin installs won't have this route. We'll need to document that in the CHANGELOG and probably release as a minor version change.

File: test/dummy/config/routes/madmin.rb [hunk 1]
```diff
@@ -9,6 +9,7 @@
   resources :habtms
   resources :teams
   root to: "dashboard#show"
+  delete "bulk/:resource", to: "bulk#destroy", as: "bulk_destroy"
```
Comment (excid3) [inline test/dummy/config/routes/madmin.rb:12]: Have you tested this with namespaced models like `ActiveStorage::Blob`? 

The form downcases which would be `activestorage::blob` and camelization will return `Activestorage::Blob` which would be incorrect?

I think we could skip underscore/camelize and can just pass the model name in to the URL directly. 

`DELETE /madmin/bulk/ActiveStorage::Blob`

These URLs should be fine and would simplify the implementation.

**Review by excid3 — CHANGES_REQUESTED** [review]

**Comment by jacobdaddario** [comment]
On it! I'll be away this weekend so I'll handle this some time next week.

File: test/dummy/config/routes/madmin.rb (hunk 1 above)
Comment (jacobdaddario) [reply test/dummy/config/routes/madmin.rb:12]: Fixed this! See commit 16dab74.

**Comment by jacobdaddario** [comment]
Hey! Resolved the above correction. Do any of the checkbox items need to be addressed? I was listening to the most recent version of the podcast, and it sounds like testing this might be a pain/we're not sure how to.

**Comment by excid3** [comment]
I haven't pulled down to try this out yet. 

On the checkboxes, I think the better option is to use `check_box_tag` instead of `form.check_box`. We don't want the hidden field to submit false that `form.check_box` generates. `check_box_tag` will only submit the IDs for the ones that are checked, which is exactly what we need.

At that point, we might not need the Bulk class either and can just use a before_action to set the objects.

**Comment by jacobdaddario** [comment]
Ah! I didn't think of that. I didn't realize that `check_box_tag` didn't add the hidden fields. I could make those changes and submit them if you'd like.

But regarding the checkbox items, I was actually referring to the tasks I put up in the first commit in the PR. Specifically I was wondering about the tests and authorization.

**Comment by excid3** [comment]
Authorization is out of scope for Madmin, so no need to worry about that. 

If you don't select any items but still submit the bulk action, it's fine if it does nothing.

We can write a controller test to submit a couple IDs for bulk delete and then if you want to do a system test we can have that test the Javascript.

**Comment by jacobdaddario** [comment]
I'll have a chance to finish this this weekend. Looks like the commit for this is starting to get messy. Probably going to cherry-pick the old commits and add them onto a branch off of an up-to-date master branch.

**Comment by jacobdaddario** [comment]
The changes weren't extensive, so I ultimately did not open a new PR for this. I'm not totally convinced that removing the bulk model was the right move. It feels like some logic relating to the data model got moved into the controller. For now I'll wait to see your thoughts, and then I'll write tests for whatever implementation we go with.

**Pull request closed without merging**